### ISSUE TITLE
Invalid Hook (use the Editor with Classes)

### ISSUE THREAD

**Issue by mqtik** [edtr-io/edtr-io#309]
Hi everyone, great library.

I'm trying to use it, but seems impossible.
![image](https://user-images.githubusercontent.com/42002579/81355977-6d3e6d00-90a6-11ea-8498-76f2c611ec28.png)

This happens after I install this library. I'm not even importing it nor using it.
If I remove this library and compile, it works fine.

I'm using this boilerplate: https://github.com/electron-react-boilerplate/electron-react-boilerplate
(so you can replicate it). And install this library on /app (there's two package.json)

I'm wondering if this is something about Webpack or Babel.
I'm really looking forward to try this library on our project!

Any clues?

**Comment by mqtik** [comment]
Ok, I've been able to resolve it, importing @edtr-io/core from the core package.json.

But I trade it for a new problem:

```
ERROR in ./node_modules/@edtr-io/core/src/store.tsx 9:24
Module parse failed: Unexpected token (9:24)
You may need an appropriate loader to handle this file type, currently no loaders are configured to process this file. See https://webpack.js.org/concepts#loaders
| } from 'react-redux'
|
> const createDispatchHook: (
|   context: React.Context<ReactReduxContextValue<State>>
| ) => // eslint-disable-next-line import/no-commonjs
```

Which's odd, because all the files on my project ends with **.tsx** format.

**Comment by inyono** [comment]
Hey @mqtik, thanks for filing the issue. Is there a chance you can share your current code as a repository? If so, I'd gladly take a look later today. Without seeing what you are currently doing, this is hard to reproduce. Might be a faulty Webpack config, might be a bug in how we currently bundle our code.

**Comment by mqtik** [comment]
Well, it seems Webpack it's working now.
But I can't use the module inside classes.

Does this library support classes or am I doing something wrong?

Thanks for your fasttreeply @inyono

**Comment by mqtik** [comment]
Yes. It is definitely a Hook/Class issue.

Is there something I/we can do about this?

**Comment by inyono** [comment]
I'm glad that you were able to sort out your Webpack issue.

Regarding your new issue: You can't use [hooks inside classes](https://reactjs.org/docs/hooks-faq.html#should-i-use-hooks-classes-or-a-mix-of-both), e.g. the following **won't work**:
```typescript
class SomeClassComponent extends React.Component {
  render() {
    const store = useScopedStore()
    return null
  }
}
```
There are a couple of things you can do to resolve / work around this limitation:
- You could usually write class components but use function components for Edtr.io specific stuff, e.g.:
  ```typescript
  // Your components that don't use React Hooks can still be classes
  class App extends React.Component {
    render() {
      // Inside your components, you can use function components that use React Hooks
      return <Editor />
    }
  }

  // Since this is a function component, you can use React Hooks here
  // Also, you may also call other class components again
  function Editor() {
    useSomeHook()
    return null
  }
  ```
  So basically, you use your existing components as before and only write "glue components" for Edtr.io as function components if you need to use some of the Edtr.io hooks.
- You could use other ways to do what the React Hooks provided by Edtr.io do. While we designed Edtr.io specifically for function components (since we use function components pretty much for everything), there should be an alternative for pretty much every hook we provide. E.g. `useStore()` can be replaced by using the React Context `EditorContext` directly.

Let me know if that helps. Furthermore: if you run into a specific issue, feel free to share your repository or some minimal code, so that I can be more specific to your use case.

**Comment by mqtik** [comment]
@inyono thanks so much for the reply.
I will try it and make it work!

Really looking forward to use this library and make it better!

I'll let you know if I run into a problem

**Comment by mqtik** [comment]
I'm pretty sure I'm doing something wrong:
```
class NewtEditor extends React.Component {
render() {
    return <EditorR />
  }
}
```

```
function EditorR() {
  //useSomeHook()
  return <Editor plugins={plugins} initialState={'test'} />
}
```

**Comment by inyono** [comment]
Hey @mqtik. Not sure where your specific issue lies. The code you are showing is mostly fine, only the `initialState` is in the wrong format.

I created a repository to show case setting up a minimal Edtr.io instance using classes (basically using create-react-app and following https://edtr.io/docs/getting-started, except that I used classes): https://github.com/inyono/edtr-io-class-minimal. Please have a look at that and compare that to what you have done so far. Relevant commit:
- [Minimal Edtr.io setup](https://github.com/inyono/edtr-io-class-minimal/commit/3233e0c69db6dc2ef2274e157258900107e6a5ab)
- [Defining a plugin](https://github.com/inyono/edtr-io-class-minimal/commit/769781d36f5b964fdc8bc4262ba0f23c1b7b8836)

**Comment by mqtik** [comment]
The repo might be private, I can't access it

**Comment by inyono** [comment]
You are right, sorry about that. It should be public now.

**Comment by inyono** [comment]
Closing this because of inactivity.
